core/trino-main/src/main/java/io/trino/execution/scheduler/ExponentialGrowthPartitionMemoryEstimator.java

```diff
@@ -34,13 +36,15 @@ public MemoryRequirements getInitialMemoryRequirements(Session session, DataSize
     }
 
     @Override
-    public MemoryRequirements getNextRetryMemoryRequirements(Session session, MemoryRequirements previousMemoryRequirements, ErrorCode errorCode)
+    public MemoryRequirements getNextRetryMemoryRequirements(Session session, MemoryRequirements previousMemoryRequirements, TaskStatus failedTaskStatus, ErrorCode errorCode)
     {
+        DataSize peakMemory = failedTaskStatus.getPeakMemoryReservation();
```

Comment:
If `peakMemory` is less than `previousMemory` and the task failed with `CLUSTER_OUT_OF_MEMORY` (simply because there were to many tasks running) do we still want to increase the memory requirement?

Reply:
This is a good question. I thought about it and I am not 100% sure. 

If we got `CLUSTER_OUT_OF_MEMORY` it means we are scheduling too much stuff on the single node. Requesting more memory for a task (even if not necessarily the task which was the root cause of the problem) works towards scheduling less on the node. 
Hence on the next try, we have a better chance that we succeed.

Theoretically, it is possible that we get `CLUSTER_OUT_OF_MEMORY` even if all the tasks on the node are still below the limit. And if we use the approach you suggest we would effectively retry everything with the same requirements. Then possibly (module random jitter) we get `CLUSTER_OUT_OF_MEMORY` once again. 

So even if it is not an exact science, and can limit throughput a bit I would keep the logic of scaling all tasks up on `COOM`. It feels to be more on the safe side to me. WDYT?